src/frequenz/sdk/model/_manager.py

```diff
@@ -0,0 +1,273 @@
+        """Get the models configuration.
+
+        Returns:
+            the model keys to model paths mapping.
+        """
+        allowed_keys = self._get_allowed_keys()
+        assert allowed_keys, "Allowed keys must be set."
+        return {key: self._build_path(key) for key in allowed_keys}
+
+
+@dataclass
+class SingleModelRepository(ModelRepository):
+    """Repository to manage a single model."""
+
+    model_path: str
+    """The path to a single model file."""
+
+    def _get_allowed_keys(self) -> list[Any]:
+        """Get the allowed keys for the single model repository.
+
+        Returns:
+            the allowed single key.
+        """
+        return ["Single"]
+
+    def _build_path(self, key: Any) -> str:
+        """Build the model path for the given key.
+
+        Args:
+            key: the key for which to build the model path.
+
+        Returns:
+            the model path.
+        """
+        return self.model_path
+
+
+@dataclass
+class DirectModelRepository(ModelRepository):
+    """Repository to manage models based on a direct configuration dictionary."""
+
+    model_config: dict[str, str]
+    """A dictionary mapping keys to the full paths of model files."""
+
+    def _get_allowed_keys(self) -> list[str]:
+        """Get the list of allowed keys for the model repository.
+
+        Returns:
+            A list of keys as specified in the model_config dictionary.
+        """
+        assert self.model_config, "Model configuration must be set."
+        return list(self.model_config.keys())
+
+    def _build_path(self, key: str) -> str:
+        """Build the model path for the given key, directly using the path from the model_config.
+
+        Args:
+            key: The key for which to build the model path, must be in the model_config.
+
+        Returns:
+            The full path to the model file associated with the given key.
+
+        Raises:
+            KeyError: If the given key is not found in the model_config.
+        """
+        assert self.model_config, "Model configuration must be set."
+        if key not in self.model_config:
+            raise KeyError(f"Key '{key}' not found in the model configuration.")
+        return self.model_config[key]
+
+
+@dataclass
+class _Model(Generic[T]):
+    """Represent a machine learning model."""
+
+    data: T
+    """The machine learning model."""
+
+    path: str
+    """The path to the model file."""
+
+
+class ModelManager(Generic[T]):
+    """Manage machine learning models.
+
+    The model manager acts as a central hub for retrieving models, reloading
+    models, and monitoring model paths for changes.
+    """
+
+    def __init__(self, model_repository: ModelRepository) -> None:
+        """Initialize the model manager.
+
+        Args:
+            model_repository: a model repository that defines the mapping
+                between model keys and paths.
+        """
+        self._models: dict[Any, _Model[T]] = {}
```

Comment:
Here we could add another parameter, like `U`, for the key type, or use `object` if we want to make it an opaque type. The difference between using `object` and `Any` is `Any` will pass any type checking, like `key.i_dont_exist`, while object will raise an error. Since the model repository has a typed key, I think it makes more sense to go with the extra type parameter.
```suggestion
        self._models: dict[U, _Model[T]] = {}
```